File: src/esquire/ReactiveList.kt
package esquire


class ReactiveListNode<TElement>(private val list: ReactiveList<TElement>,
                                 value: TElement): LinkedListNode<TElement>(value) {
    val onRemove = EventDispatcher<Unit>()

    override fun removeImpl() {
        if(removed)
            return
        super.removeImpl()
        list.size.value -= 1
        list.onRemove(value)
        onRemove()
    }
}

class ReactiveList<TElement> {
    private val head = ReactiveListNode<TElement>(this, null)
    private val tail = ReactiveListNode<TElement>(this, null)

    init {
        head.next = tail
    }

    val size = Reactant(0)
    val onRemove = EventDispatcher<TElement>()
    val onAdd = EventDispatcher<TElement>()
    val onAddNode = EventDispatcher<ReactiveListNode<TElement>>()

    fun addFirst(element: TElement): ReactiveListNode<TElement> {
        val node = ReactiveListNode(this, element)
        head.insertAfter(node)
        size.value += 1
        onAddNode(node)
        onAdd(element)
        return node
    }

    fun addLast(element: TElement): ReactiveListNode<TElement> {
        val node = ReactiveListNode(this, element)
        tail.insertBefore(node)
        size.value += 1
        onAddNode(node)
        onAdd(element)
        return node
    }

    val empty: Boolean
        get() = head.next === tail

    fun clear() {
        head.next = tail
    }

    fun forEach(callback: (TElement)->Unit) = forEachNode { callback(it.value) }

    fun forEachNode(callback: (ReactiveListNode<TElement>)->Unit) {
        var node = head
        while(true) {
            node = node.next!! as ReactiveListNode<TElement>
            if(node === tail)
                return
            if(node.removed)
                continue
            callback(node)
        }
    }

    fun onAddInvoke(callback: (TElement)->Unit): Registration {
        forEach(callback)
        return onAdd.listen(callback)
    }

    fun onAddNodeInvoke(callback: (ReactiveListNode<TElement>)->Unit): Registration {
        forEachNode(callback)
        return onAddNode.listen(callback)
    }
}

File: src/reactivelist.ts
'use strict'

if(typeof require !== 'undefined') {
    var LinkedList_:any = require('./linkedlist')
    var LinkedListNode_:any = require('./linkedlistnode')
    var Reactant_:any = require('./reactant')
    var EventDispatcher_:any = require('./eventdispatcher')
}
else {
    var LinkedList_:any = LinkedList
    var LinkedListNode_:any = LinkedListNode
    var Reactant_:any = Reactant
    var EventDispatcher_:any = EventDispatcher
}

class ReactiveListNode {
    removed
    base_remove
    onRemove
    
    constructor(private list, public value?) {
        LinkedListNode_.call(this, value)
        this.onRemove = new EventDispatcher_()
    }
    remove() {
        if(this.removed)
            return
        this.base_remove()
        this.list.size.value -= 1
        this.list.onRemove.proc(this.value)
        this.onRemove.proc()
    }
}
ReactiveListNode.prototype['__proto__'] = LinkedListNode_.prototype

class ReactiveList {
    size
    head
    tail
    onAdd
    onAddNode
    onRemove
    forEach
    forEachNode

    constructor() {
        LinkedList_.call(this)
        Object.defineProperties(this, {
            size: {value: new Reactant_(0)},
            onAdd: {value: new EventDispatcher_()},
            onAddNode: {value: new EventDispatcher_()},
            onRemove: {value: new EventDispatcher_()}
        })
    }
    
    addFirst(element) {
        var node = new ReactiveListNode(this, element)
        this.head.insertAfter(node)
        this.size.value += 1
        this.onAddNode.proc(node)
        this.onAdd.proc(element)
        return node
    }
    addLast(element) {
        var node = new ReactiveListNode(this, element)
        this.tail.insertBefore(node)
        this.size.value += 1
        this.onAddNode.proc(node)
        this.onAdd.proc(element)
        return node
    }
    
    onAdd_pc(callback) {
        this.forEach(callback)
        return this.onAdd.listen(callback)
    }
    onAddNode_pc(callback) {
        this.forEachNode(callback)
        return this.onAddNode.listen(callback)
    }
}
ReactiveList.prototype['__proto__'] = LinkedList_.prototype

if(typeof module !== 'undefined')
    module.exports = ReactiveList
